### DISCUSSION TITLE
Load environment variables script only when executing tasks.

### DISCUSSION THREAD

**Discussion by uptickmetachu** [jdx/mise#3550] [Ideas]
We can currently set `_.env` to a shell script and read the exported variables into the mise environment.

It would be nice to have a `_.env` scoped to tasks and only executed/templated when tasks are run. This would constant execution during regular shell usage.

**Comment by jdx** [comment]
I think we could have a `for="tasks"` option on `env._.source` or something. We could also probably support having `tasks.<TASK>.env._.source` on the task env vars too

**Comment by egnor** [comment]
Sourcing the script (and running a bash subprocess, etc) _every time_ there's a new shell prompt does seem a little bit much, though technically it's the only way to be entirely correct. Direnv just runs its stuff on entry to the environment and honestly that seems good for 99% of cases. Checking for modifications on the script being sourced would get you to 99.9% I think?

**Comment by uptickmetachu** [comment]
As you said sourcing the script every time is the only way to be 100% correct. Something simple like
`export TIME=echo$(date)` will be wrong if we only check the output after script modification.

The use case I have is setting up environment variables for a heavily parallelised build process. I could source the file before running the tasks but I would prefer to keep it consistent with having mise as the entrypoint.

I like the idea of `tasks.env._.source` and it does seem cleaner than adding a `for` option on `env._.source`. It was also the first I checked for in the documentation.

**Comment by jdx** [comment]
FYI there does seem to currently be a bug in "short-circuiting" logic that prevents running scripts if nothing has been changed: https://github.com/jdx/mise/issues/1617

**Comment by jdx** [comment]
I'm a little curious how this behaves now with the hooks changes in, this may be solved now. If not, a hacky solution would be a hook on enter which just calls `touch .env`—though I don't know if I would call that a "solution" to this. Might be.

**Comment by Marukome0743** [comment]
This is now supported by [PR #5638](https://github.com/jdx/mise/pull/5638).

Task-specific environment directives can be defined under `[tasks.<name>.env]`, including `_.source`, `_.file`, and `_.path`. For example:

```toml
[tasks.build]
run = "./build.sh"

[tasks.build.env]
_.source = "./build-env.sh"
```

In this example, `build-env.sh` is resolved as part of the `build` task’s environment. It does not become part of the normal shell activation or prompt-time `hook-env` environment, so it avoids repeatedly sourcing the script during regular shell usage.

Different tasks can also use different source scripts or environment files. This implements the `tasks.<TASK>.env._.source` approach discussed above, so the `touch .env` workaround should no longer be necessary on current mise versions.

*AI-assisted — Tool: Codex; model: OpenAI/GPT-5; version: unavailable.*
